connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -279,15 +291,61 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
+                .filter(offsetEntry -> offsetEntry.getValue() == null)
+                .map(Map.Entry::getKey).collect(Collectors.toList());
+
+        Map<ByteBuffer, ByteBuffer> tombstoneOffsets = new HashMap<>();
+        for (ByteBuffer partition : tombstoneOffsetPartitions) {
+            tombstoneOffsets.put(partition, null);
+        }
+        Map<ByteBuffer, ByteBuffer> regularOffsets = values.entrySet().stream()
+                .filter(offsetEntry -> offsetEntry.getValue() != null)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        // If the supplied offsets contain tombstone values, then tombstone offsets are extracted first and
+        // we first write to the secondary store in a synchronous manner and then to primary store.
+        // This is because, if a tombstone offset is successfully written to the per-connector offsets topic,
+        // but cannot be written to the global offsets topic, then the global offsets topic will still contain that
+        // source offset, but the per-connector topic will not. Due to the fallback-on-global logic used by the worker,
+        // if a task requests offsets for one of the tombstoned partitions, the worker will provide it with the
+        // offsets present in the global offsets topic, instead of indicating to the task that no offsets can be found.
+        KafkaFutureImpl<Void> write = new KafkaFutureImpl<>();
+        write.thenApply(v -> {
+            FutureCallback<Void> secondaryWriteFuture = new FutureCallback<>();
+            if (secondaryStore != null && !tombstoneOffsets.isEmpty()) {
+                secondaryStore.set(tombstoneOffsets, secondaryWriteFuture);
+                try {
+                    if (exactlyOnce) {
+                        secondaryWriteFuture.get();
+                    } else {
+                        secondaryWriteFuture.get(offsetFlushTimeoutMs, TimeUnit.MILLISECONDS);
+                    }
+                } catch (InterruptedException e) {
+                    log.warn("{} Flush of tombstone(s) offsets to secondary store interrupted, cancelling", this);
+                } catch (ExecutionException e) {
+                    log.error("{} Flush of tombstone(s) offsets to secondary store threw an unexpected exception: ", this, e.getCause());
+                } catch (TimeoutException e) {
+                    log.error("{} Timed out waiting to flush tombstone(s) offsets to secondary store ", this);
+                } catch (Exception e) {
+                    log.error("{} Got Exception when trying to flush tombstone(s) offsets to secondary store", this, e);
+                }
+            }
+            return secondaryWriteFuture;
+        });
+
+        // Need to explicitly invoke complete as otherwise the subsequent write to primary store doesn't work since it's
+        // waiting for this future to complete.
+        write.complete(null);
```

Comment:
I am still deliberating is this is the right thing to do or not. But as commented, w/o this it never goes to the below `thenApply` block.

Reply:
This doesn't actually change the semantics of the method from when it was last reviewed, since it still causes a synchronous wait on the first flush to the secondary store.

Reply:
Yeah that’s what I was thinking as well. Ideally the second function should be invoked once the first one completes but the tests throw timeout exception if I don’t have this call.

Reply:
I removed that `complete` call. To get around the need for it's usage, I created a dummy completed future [here](https://github.com/apache/kafka/pull/13801/files#diff-0014cfa4c32bfde215cb10bee987e997d5182815fcf1a1245539375519f83f3dR312) upfront which triggers the subsequent `thenApply` calls which are set up on a conditional basis.